lib/Sema/CSSolver.cpp

```diff
@@ -1964,3 +1966,41 @@ bool DisjunctionChoice::isSymmetricOperator() const {
       paramList->get(1)->getInterfaceType()->getWithoutSpecifierType();
   return firstType->isEqual(secondType);
 }
+
+void DisjunctionChoice::propagateConversionInfo() const {
+  if (!CS->isExplicitConversionConstraint(Disjunction))
+    return;
+
+  auto LHS = Choice->getFirstType();
+  auto typeVar = LHS->getAs<TypeVariableType>();
+  if (!typeVar || typeVar->getImpl().hasRepresentativeOrFixed())
+    return;
+
+  auto bindings = CS->getPotentialBindings(typeVar);
+  if (bindings.InvolvesTypeVariables || bindings.Bindings.size() != 1)
+    return;
+
+  auto conversionType = bindings.Bindings[0].BindingType;
+  SmallVector<Constraint *, 4> constraints;
+  CS->CG.gatherConstraints(typeVar, constraints,
+                           ConstraintGraph::GatheringKind::EquivalenceClass);
+
+  bool viableForBinding = true;
+  for (auto adjacent : constraints) {
+    switch (adjacent->getKind()) {
+    case ConstraintKind::Conversion:
+    case ConstraintKind::Defaultable:
+    case ConstraintKind::ConformsTo:
+    case ConstraintKind::LiteralConformsTo:
+      break;
+
+    default:
+      viableForBinding = false;
+      break;
+    }
+  }
+
+  if (viableForBinding)
+    CS->addConstraint(ConstraintKind::Bind, typeVar, conversionType,
+                      Choice->getLocator());
+}
```

Comment:
Okay, so now seeing where this gets called I think my main question is whether this should be generalized and incorporated into getPotentialBindings?

For example I've noticed that it seems like it's often beneficial to extend a solution from the most recent decision we've made, e.g. selecting a something from a disjunction and then examining the types for the constraint we've selected as is the case here, or selecting a type for one side of a constraint and then moving on to look at the other side of the constraint.

Reply:
Good point but the problem is that, as far as a recall, locator for conversion choice isn't going to have any information related to coercion disjunction which makes it hard to prioritize in `getPotentialBindings`, but I think as soon as we get our story straight about when/how to try bindings it could be moved there.